docs: Provide more structured API docs for Workflow API (#3252)
dapr/docs: daprdocs/content/en/reference/api/workflow_api.md

description: "Detailed documentation on the workflow API"
weight: 900
---

Dapr provides users with the ability to interact with workflows and comes with a built-in `dapr` component.

## Start workflow request

> Start a workflow instance with the given name and instance id.

Reviewer: ```suggestion
Start a workflow instance with the given name and instance ID.
```

`workflowName` | Identify the workflow type
`instanceId` | Unique value created for each run of a specific workflow

### Request Contents

In the request you can pass along headers:

```json
{
  "input": {

>     // arguments to pass to the workflow

Reviewer: Calling it "arguments" (plural) isn't quite correct since workflows can only take a single argument. Also, the argument doesn't necessarily have to be a JSON object, which seems to be implied by the `{...}` here. It can be any valid JSON data, including objects, strings, numbers, or arrays.

Author: Updated with a proposal based on what you shared - Thoughts?

Reviewer: Yeah, I like the update you provided.

  }
}
```

### HTTP Response codes

Code | Description
---- | -----------
`202`  | Accepted
`400`  | Request was malformed
`500`  | Request formatted correctly, error in dapr code or underlying component

### Response Contents

The API call will provide a response similar to this:

```json
{
  "WFInfo": {
    "instance_id": "SampleWorkflow"
  }
}
```

## Terminate workflow request

Terminate a running workflow instance with the given name and instance id.

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceId>/terminate
```

### URL parameters

Parameter | Description
--------- | -----------
`workflowComponentName` | Current default is `dapr` for Dapr Workflows
`workflowName` | Identify the workflow type
`instanceId` | Unique value created for each run of a specific workflow

### HTTP Response codes

Code | Description
---- | -----------
`202`  | Accepted
`400`  | Request was malformed
`500`  | Request formatted correctly, error in dapr code or underlying component

### Response Contents

The API call will provide a response similar to this:

```bash

> HTTP/1.1 202 Accepted

Reviewer: I'm just noticing that the `Content-Type` and `Content-Length` fields in the response looks wrong. Would you mind removing these for us as part of this PR?

`workflowName` | Identify the workflow type
`instanceId` | Unique value created for each run of a specific workflow

> ### Request Contents

Reviewer: ```suggestion
### Request content
```

> In the request you can pass along headers:

Reviewer: We don't accept any specific headers. Did you mean to say something about the workflow input here?

Author: I guess so, I forgot 🤷‍♂️

What about this?

```suggestion
In the request you can pass along relevant input information that will be passed to the workflow:
```

Reviewer: LGTM!

```json
{
  "input": // argument(s) to pass to the workflow which can be any valid JSON data type (such as objects, strings, numbers, arrays, etc.)
}
```

> ### HTTP Response codes

Reviewer: ```suggestion
### HTTP response codes
```

Code | Description
---- | -----------
`202`  | Accepted
`400`  | Request was malformed
`500`  | Request formatted correctly, error in dapr code or underlying component

> ### Response Contents

Reviewer: ```suggestion
### Response content
```

The API call will provide a response similar to this:

```json
{
  "WFInfo": {
    "instance_id": "SampleWorkflow"
  }
}
```

## Terminate workflow request

> Terminate a running workflow instance with the given name and instance id.

Reviewer: ```suggestion
Terminate a running workflow instance with the given name and instance ID.
```

```bash
POST http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<instanceId>/terminate
```

### URL parameters

Parameter | Description
--------- | -----------
`workflowComponentName` | Current default is `dapr` for Dapr Workflows
`workflowName` | Identify the workflow type
`instanceId` | Unique value created for each run of a specific workflow

> ### HTTP Response codes

Reviewer: ```suggestion
### HTTP response codes
```

Code | Description
---- | -----------
`202`  | Accepted
`400`  | Request was malformed
`500`  | Request formatted correctly, error in dapr code or underlying component

> ### Response Contents

Reviewer: ```suggestion
### Response content
```

The API call will provide a response similar to this:

```bash
HTTP/1.1 202 Accepted
Server: fasthttp
Date: Thu, 12 Jan 2023 21:31:16 GMT
Traceparent: 00-e3dedffedbeb9efbde9fbed3f8e2d8-5f38960d43d24e98-01
Connection: close 
```

### Get workflow request

Get information about a given workflow instance.

```bash
GET http://localhost:3500/v1.0-alpha1/workflows/<workflowComponentName>/<workflowName>/<instanceId>
```

### URL parameters

Parameter | Description
--------- | -----------
`workflowComponentName` | Current default is `dapr` for Dapr Workflows
`workflowName` | Identify the workflow type
`instanceId` | Unique value created for each run of a specific workflow

> ### HTTP Response codes

Reviewer: ```suggestion
### HTTP response codes
```

Code | Description
---- | -----------
`202`  | Accepted
`400`  | Request was malformed
`500`  | Request formatted correctly, error in dapr code or underlying component

> ### Response Contents

Reviewer: ```suggestion
### Response content
```

 }
```

## Headers

> As part of the start HTTP request, the caller can optionally include one or more `dapr-workflow-metadata` HTTP request headers. The format of the header value is a list of `{key}={value}` values, similar to the format for HTTP cookie request headers. These key/value pairs are saved in the workflow instance’s metadata and can be made available for search (in cases where the workflow implementation supports this kind of search).

Reviewer: Can we remove this whole `## Headers` section? It's not correct. We have no such feature today.

Author: That is what I figured but kept it around, will remove it.

Author: Updated

## Component format

A Dapr `workflow.yaml` component file has the following structure:

```yaml
apiVersion: dapr.io/v1alpha1
kind: Component
metadata:
  name: <NAME>
spec:
  type: workflow.<TYPE>
  version: v1.0-alpha1
  metadata:
  - name: <NAME>
    value: <VALUE>
 ```

| Setting | Description |
| ------- | ----------- |
| `metadata.name` | The name of the workflow component. |
| `spec/metadata` | Additional metadata parameters specified by workflow component |

> However, Dapr comes with a built-in `dapr` workflow component  that is built on Dapr Actors.

Reviewer: ```suggestion
However, Dapr comes with a built-in `dapr` workflow component that is built on Dapr Actors. No component file is required to use the built-in Dapr workflow component.
```

Author: Updated
